File: app/src/main/java/xyz/zpayh/myadapter/MultiItemActivity.java
package xyz.zpayh.myadapter;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import xyz.zpayh.adapter.IMultiItem;
import xyz.zpayh.adapter.OnItemLongClickListener;
import xyz.zpayh.myadapter.adapter.MyMultiAdapter;
import xyz.zpayh.myadapter.data.Image;
import xyz.zpayh.myadapter.data.Text;

public class MultiItemActivity extends AppCompatActivity {

    private RecyclerView mRecyclerView;
    private MyMultiAdapter mAdapter;
    private List<IMultiItem> mData;

    private GridLayoutManager mGridLayoutManager;
    private StaggeredGridLayoutManager mStaggeredGridLayoutManager;

    private boolean mIsGrid;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_multi_item);
        initView();
    }

    private void initView() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mGridLayoutManager = new GridLayoutManager(this,3);
        mStaggeredGridLayoutManager = new StaggeredGridLayoutManager(2,StaggeredGridLayoutManager.VERTICAL);

        final SwipeRefreshLayout refreshLayout = (SwipeRefreshLayout) findViewById(R.id.refresh);

        mRecyclerView = (RecyclerView) findViewById(R.id.list);
        mRecyclerView.setLayoutManager(mGridLayoutManager);
        mIsGrid = true;
        mAdapter = new MyMultiAdapter();
        mRecyclerView.setAdapter(mAdapter);

        mAdapter.setOnItemLongClickListener(new OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(@NonNull View view, final int adapterPosition) {
                new AlertDialog.Builder(MultiItemActivity.this)
                        .setTitle("是否删除第"+adapterPosition+"项")
                        .setPositiveButton("确定", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                mAdapter.removeData(adapterPosition);
                            }
                        })
                        .setNegativeButton("取消", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {

                            }
                        })
                        .create()
                        .show();
                return true;
            }
        });
        mData = new ArrayList<>();
        initGridData();

        refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refreshLayout.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        mAdapter.setData(mData);
                        refreshLayout.setRefreshing(false);
                    }
                },500);
            }
        });
    }

    private void initGridData() {
        mData.clear();
        String[] list = getResources().getStringArray(R.array.list);
        for (int i = 0; i < list.length; i++) {
            mData.add(new Text(list[i],i%3+1));
        }

        int width = getResources().getDisplayMetrics().widthPixels / mGridLayoutManager.getSpanCount();

        mData.add(2,new Image(R.drawable.girl,width));
        mData.add(2,new Image(R.drawable.girl,width,2));
        mData.add(2,new Image(R.drawable.easter_eggs,width,2));
        mData.add(2,new Image(R.drawable.sunset,width,2));
        mData.add(4,new Image(R.drawable.adult,width,3));
        mData.add(7,new Image(R.drawable.sunset,width));

        mAdapter.setData(mData);
    }

    private void initStaggeredGridData() {
        mData.clear();
        String[] list = getResources().getStringArray(R.array.list);
        for (int i = 0; i < list.length; i++) {
            mData.add(new Text(list[i],i%3+1));
        }

        int width = getResources().getDisplayMetrics().widthPixels / mStaggeredGridLayoutManager.getSpanCount();

        mData.add(2,new Image(R.drawable.girl,width));
        mData.add(2,new Image(R.drawable.girl,width));
        mData.add(2,new Image(R.drawable.easter_eggs,width));
        mData.add(2,new Image(R.drawable.adult,width));
        mData.add(2,new Image(R.drawable.sunset,width));
        mData.add(2,new Image(R.drawable.easter_eggs,width));
        mData.add(2,new Image(R.drawable.adult,width));
        mData.add(7,new Image(R.drawable.sunset,width));

        mAdapter.setData(mData);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_grid:
                if (mIsGrid){
                    return true;
                }
                mIsGrid = true;
                mRecyclerView.setLayoutManager(mGridLayoutManager);
                mAdapter.onAttachedToRecyclerView(mRecyclerView);
                initGridData();
                return true;
            case R.id.action_staggered:
                if (!mIsGrid){
                    return true;
                }
                mIsGrid = false;
                mRecyclerView.setLayoutManager(mStaggeredGridLayoutManager);
                initStaggeredGridData();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.multiitem_menu, menu);
        return true;
    }
}

File: sample-kotlin/src/main/java/cn/sherlockzp/sample/MultiItemActivity.kt
package cn.sherlockzp.sample

import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.Menu
import android.view.MenuItem
import cn.sherlockzp.adapter.BaseExpandableAdapter
import cn.sherlockzp.adapter.BaseMultiAdapter
import cn.sherlockzp.adapter.IMultiItem
import cn.sherlockzp.sample.model.Image
import cn.sherlockzp.sample.model.Text

import kotlinx.android.synthetic.main.activity_multi_item.*

class MultiItemActivity : AppCompatActivity() {

    private val gridLayoutManager by lazy {
        GridLayoutManager(this,3)
    }

    private val staggeredGridLayoutManager by lazy {
        StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
    }

    private val adapter = BaseMultiAdapter() //BaseExpandableAdapter()
    private val data = ArrayList<IMultiItem>()

    private var isGrid = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_multi_item)

        initView()
    }

    private fun initView() {
        rv_list_multi.layoutManager = gridLayoutManager
        rv_list_multi.adapter = adapter

        adapter.setOnItemLongClickListener { _, adapterPosition ->
            AlertDialog.Builder(this)
                    .setTitle("是否删除第${adapterPosition}项")
                    .setPositiveButton("确定"){
                        _, _ ->
                        adapter.removeDataAt(adapterPosition)
                    }
                    .setNegativeButton("取消",null)
                    .create()
                    .show()
            return@setOnItemLongClickListener true
        }

        initGridData()

        sr_refresh.setOnRefreshListener {
            sr_refresh.postDelayed({
                adapter.setData(data)
                sr_refresh.isRefreshing = false
            },500)
        }
    }

    private fun initGridData(){
        data.clear()
        val list = resources.getStringArray(R.array.list)
        for (i in list.indices) {
            data.add(Text(list[i], i % 3 + 1))
        }

        val width = resources.displayMetrics.widthPixels / gridLayoutManager.spanCount

        data.add(2,Image(R.drawable.girl,width * 2,2))
        data.add(3,Image(R.drawable.easter_eggs,width))
        data.add(4,Image(R.drawable.adult,width * 3,3))

        data.add(7,Image(R.drawable.girl1,width * 3,3))
        data.add(7,Image(R.drawable.sunset,width))
        data.add(7,Image(R.drawable.horses,width))

        adapter.setData(data)
    }

    private fun initStaggeredGridData(){
        data.clear()
        val list = resources.getStringArray(R.array.list)
        for (i in list.indices) {
            data.add(Text(list[i], i % 3 + 1))
        }

        val width = resources.displayMetrics.widthPixels / staggeredGridLayoutManager.spanCount

        data.add(2,Image(R.drawable.girl,width))
        data.add(2,Image(R.drawable.girl1,width))
        data.add(2,Image(R.drawable.easter_eggs,width))
        data.add(2,Image(R.drawable.adult,width))
        data.add(2,Image(R.drawable.sunset,width))
        data.add(2,Image(R.drawable.fashion,width))
        data.add(2,Image(R.drawable.notes,width))
        data.add(7,Image(R.drawable.girl2,width))

        adapter.setData(data)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when(item?.itemId) {
            R.id.action_grid -> {
                if (isGrid) {
                    return true
                }
                isGrid = true
                rv_list_multi.layoutManager = gridLayoutManager
                adapter.onAttachedToRecyclerView(rv_list_multi)
                initGridData()
                return true
            }
            R.id.action_staggered -> {
                if (!isGrid) {
                    return true
                }
                isGrid = false
                rv_list_multi.layoutManager = staggeredGridLayoutManager
                adapter.onAttachedToRecyclerView(rv_list_multi)
                initStaggeredGridData()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.multiitem_menu, menu)
        return true
    }
}
